staging/src/k8s.io/apiserver/pkg/server/options/serving.go

```diff
@@ -222,47 +222,85 @@ func (s *SecureServingOptions) ApplyTo(config **server.SecureServingInfo) error
 	}
 	c := *config
 
-	serverCertFile, serverKeyFile := s.ServerCert.CertKey.CertFile, s.ServerCert.CertKey.KeyFile
-	// load main cert
-	if len(serverCertFile) != 0 || len(serverKeyFile) != 0 {
-		tlsCert, err := tls.LoadX509KeyPair(serverCertFile, serverKeyFile)
-		if err != nil {
-			return fmt.Errorf("unable to load server certificate: %v", err)
-		}
-		c.Cert = &tlsCert
-	} else if s.ServerCert.GeneratedCert != nil {
-		c.Cert = s.ServerCert.GeneratedCert
-	}
-
+	var cipherSuites []uint16
+	var err error
 	if len(s.CipherSuites) != 0 {
-		cipherSuites, err := cliflag.TLSCipherSuites(s.CipherSuites)
+		cipherSuites, err = cliflag.TLSCipherSuites(s.CipherSuites)
 		if err != nil {
 			return err
 		}
-		c.CipherSuites = cipherSuites
 	}
 
-	var err error
-	c.MinTLSVersion, err = cliflag.TLSVersion(s.MinTLSVersion)
+	minTLSVersion, err := cliflag.TLSVersion(s.MinTLSVersion)
 	if err != nil {
 		return err
 	}
 
-	// load SNI certs
-	namedTLSCerts := make([]server.NamedTLSCert, 0, len(s.SNICertKeys))
-	for _, nck := range s.SNICertKeys {
-		tlsCert, err := tls.LoadX509KeyPair(nck.CertFile, nck.KeyFile)
-		namedTLSCerts = append(namedTLSCerts, server.NamedTLSCert{
-			TLSCert: tlsCert,
-			Names:   nck.Names,
-		})
+	c.GetConfigForClient = func(_ *tls.ClientHelloInfo) (*tls.Config, error) {
+		config := &tls.Config{
+			// Can't use SSLv3 because of POODLE and BEAST
+			// Can't use TLSv1.0 because of POODLE and BEAST using CBC cipher
+			// Can't use TLSv1.1 because of RC4 cipher usage
+			MinVersion: tls.VersionTLS12,
+			// enable HTTP2 for go's 1.7 HTTP Server
+			NextProtos: []string{"h2", "http/1.1"},
+		}
+
+		if minTLSVersion > 0 {
+			config.MinVersion = minTLSVersion
+		}
+		if len(cipherSuites) > 0 {
+			config.CipherSuites = cipherSuites
+		}
+
+		// load main cert
+		serverCertFile, serverKeyFile := s.ServerCert.CertKey.CertFile, s.ServerCert.CertKey.KeyFile
```

Comment:
this needs to follow a similar approach to the bearer token handling in client go:

1. at startup, synchronously load the referenced files (startup should still fail if the files cannot be read/parsed)
2. periodically (either on a time interval or potentially in response to a signal) attempt to reread the files
3. request handling uses the last successfully read values if there is an error later reading the files

Reply:
the simplest would probably be for the re-read to construct a *tls.Config and on success `Store()` it in an [atomic.Value](https://golang.org/pkg/sync/atomic/#Value), then GetConfigForClient could just `return configHolder.Load().(*tls.Config)`, without needing to lock

Reply:
@liggitt thanks for your suggestions. I have updated the branch. I kept with a time interval over a signal for now